src/libappimage/libappimage.cpp

```diff
@@ -0,0 +1,364 @@
+/**
+ * Implementation of the C interface functions
+ */
+// system
+#include <cstring>
+#include <sstream>
+#include <type_traits> //for std::underlying_type
+
+// libraries
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string.hpp>
+
+// local
+#include <XdgUtils/DesktopEntry/DesktopEntry.h>
+#include <appimage/core/AppImage.h>
+#include "utils/HashLib.h"
+#include "utils/UrlEncoder.h"
+
+#ifdef LIBAPPIMAGE_DESKTOP_INTEGRATION_ENABLED
+
+#include <appimage/desktop_integration/IntegrationManager.h>
+#endif
+
+using namespace appimage::core;
+
+namespace bf = boost::filesystem;
+
+extern "C" {
+
+/* Check if a file is an AppImage. Returns the image type if it is, or -1 if it isn't */
+int appimage_get_type(const char* path, bool verbose) {
+    typedef std::underlying_type<AppImageFormat>::type utype;
+
+    try {
+        AppImage appImage(path);
+        return static_cast<utype>(appImage.getFormat());
+    } catch (const std::runtime_error& err) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    return static_cast<utype>(AppImageFormat::INVALID);
+}
+
+char** appimage_list_files(const char* path) {
+    char** result = nullptr;
+    try {
+        AppImage appImage(path);
+
+        std::vector<std::string> files;
+        for (auto itr = appImage.files(); itr != itr.end(); ++itr)
+            if (!(*itr).empty())
+                files.emplace_back(*itr);
+
+
+        result = static_cast<char**>(malloc(sizeof(char*) * (files.size() + 1)));
+        for (int i = 0; i < files.size(); i++)
+            result[i] = strdup(files[i].c_str());
+
+        result[files.size()] = nullptr;
+
+        return result;
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    // Create empty string list
+    result = static_cast<char**>(malloc(sizeof(char*)));
```

Comment:
A method to create a new char buffer for a given vector in our "string list" format is needed more than once (see also line 62), so we should make a function. To produce an empty list like here, you could then just pass an empty vector (or specialize the method w/o arguments, although that's again redundant then).